test/unit/prof_threshold.c

```diff
@@ -0,0 +1,103 @@
+#include "test/jemalloc_test.h"
+
+// Test config (set in reset_test_config)
+const int ALLOC_ITERATIONS_IN_THRESHOLD = 10;
+uint64_t threshold_bytes = 0;
+uint64_t chunk_size = 0;
+
+// Test globals for calblack
+uint64_t hook_calls = 0;
+uint64_t last_peak = 0;
+uint64_t last_alloc = 0;
+uint64_t alloc_baseline = 0;
+
+void
+mock_prof_threshold_hook(uint64_t alloc, uint64_t dealloc, uint64_t peak) {
+	hook_calls++;
+	last_peak = peak;
+	last_alloc = alloc;
+}
+
+/* Need the do_write flag because NULL is a valid to_write value. */
+static void
+read_write_prof_threshold_hook(prof_threshold_hook_t *to_read, bool do_write,
+    prof_threshold_hook_t to_write) {
+	size_t hook_sz = sizeof(prof_threshold_hook_t);
+	expect_d_eq(mallctl("experimental.hooks.prof_threshold",
+	    (void *)to_read, &hook_sz, do_write ? &to_write : NULL, hook_sz), 0,
+	    "Unexpected prof_threshold_hook mallctl failure");
+}
+
+static void
+write_prof_threshold_hook(prof_threshold_hook_t new_hook) {
+	read_write_prof_threshold_hook(NULL, true, new_hook);
+}
+
+static prof_threshold_hook_t
+read_prof_threshold_hook() {
+	prof_threshold_hook_t hook;
+	read_write_prof_threshold_hook(&hook, false, NULL);
+	return hook;
+}
+
+static void reset_test_config() {
+	hook_calls = 0;
+	last_peak = 0;
+	alloc_baseline = last_alloc; // We run the test multiple times
+	last_alloc = 0;
+	threshold_bytes = 1 << opt_experimental_lg_prof_threshold;
+	chunk_size = threshold_bytes / ALLOC_ITERATIONS_IN_THRESHOLD;
+}
+
+static void expectThresholdCalls(int calls) {
+	expect_zu_eq(hook_calls, calls, "Hook called the right amount of times");
+	expect_true(last_peak < chunk_size * 2, "We allocate chunk_size at a time");
+	expect_true(last_alloc >= threshold_bytes * calls + alloc_baseline, "Crosses");
```

Comment:
For a clearer error message purposes.

```suggestion
	expect_u64_lt(last_peak, chunk_size * 2, "We allocate chunk_size at a time");
	expect_u64_ge(last_alloc, threshold_bytes * calls + alloc_baseline, "Crosses");
```